File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTSource.java
package com.mapbox.rctmgl.components.styles.sources;

import com.facebook.react.bridge.ReadableMap;
import com.mapbox.geojson.Feature;
import com.mapbox.rctmgl.components.AbstractMapFeature;
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView;
import com.mapbox.rctmgl.components.styles.layers.RCTLayer;

import com.mapbox.rctmgl.utils.LatLng;
import com.mapbox.rctmgl.utils.Logger;

import com.mapbox.maps.extension.style.sources.Source;
import com.mapbox.maps.MapboxMap;
import com.mapbox.maps.Style;
import com.mapbox.maps.extension.style.sources.SourceUtils;

import android.content.Context;
import android.graphics.PointF;
import android.view.View;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class RCTSource<T extends Source> extends AbstractMapFeature {
    public static final String DEFAULT_ID = "composite";
    public static final String LOG_TAG = "RCTSource";

    protected RCTMGLMapView mMapView;
    protected MapboxMap mMap;

    protected String mID;
    protected T mSource;
    protected boolean mHasPressListener;
    protected Map<String, Double> mTouchHitbox;

    protected List<AbstractSourceConsumer> mLayers;
    private List<AbstractSourceConsumer> mQueuedLayers;

    public RCTSource(Context context) {
        super(context);
        mLayers = new ArrayList<>();
        mQueuedLayers = new ArrayList<>();
    }

    public String getID() {
        return mID;
    }
    
    private T getSourceAs(Style style, String id) {
        Source result = SourceUtils.getSource(style, mID);

        try {
            return (T)result;
        } catch (ClassCastException exception) {
            return null;
        }
    }

    protected void addLayerToMap(AbstractSourceConsumer layer, int childPosition) {
        if (mMapView == null || layer == null) {
            return;
        }

        layer.addToMap(mMapView);
        if (!mLayers.contains(layer)) {
            mLayers.add(childPosition, layer);
        }
    }

    public boolean hasPressListener() {
        return mHasPressListener;
    }

    public void setHasPressListener (boolean hasPressListener) {
        mHasPressListener = hasPressListener;
    }

    public void setHitbox(ReadableMap map) {
        Map<String, Double> hitbox = new HashMap<>();
        hitbox.put("width", map.getDouble("width"));
        hitbox.put("height", map.getDouble("height"));
        mTouchHitbox = hitbox;
    }

    public void setID(String id) {
        mID = id;
    }

    public int getLayerCount () {
        int totalCount = 0;

        if (mQueuedLayers != null) {
            totalCount = mQueuedLayers.size();
        }

        totalCount += mLayers.size();
        return totalCount;
    }

    @Override
    public void addToMap(RCTMGLMapView mapView) {
        mMapView = mapView;
        mMap = mapView.getMapboxMap();
        mMap.getStyle(new Style.OnStyleLoaded() {
            public void onStyleLoaded(@NonNull Style style) {
                T existingSource = getSourceAs(style, mID);
                if (existingSource != null) {
                    mSource = existingSource;
                } else {
                    mSource = makeSource();
                    SourceUtils.addSource(style, mSource);
                }

                if (mQueuedLayers != null && mQueuedLayers.size() > 0) { // first load
                    for (int i = 0; i < mQueuedLayers.size(); i++) {
                        addLayerToMap(mQueuedLayers.get(i), i);
                    }
                    mQueuedLayers = null;
                } else if (mLayers.size() > 0) { // handles the case of switching style url, but keeping layers on map
                    for (int i = 0; i < mLayers.size(); i++) {
                        addLayerToMap(mLayers.get(i), i);
                    }
                }
            }
        });


    }

    @Override
    public void removeFromMap(RCTMGLMapView mapView) {
        if (mLayers.size() > 0) {
            for (int i = 0; i < mLayers.size(); i++) {
                AbstractSourceConsumer layer = mLayers.get(i);
                layer.removeFromMap(mMapView);
            }
        }
        if (mQueuedLayers != null) {
            mQueuedLayers.clear();
        }
        if (mMap != null && mSource != null && mMap.getStyle() != null) {
            try {
                mMap.getStyle().removeStyleSource(mID);
            } catch (Throwable ex) {
                Logger.w(LOG_TAG, String.format("RCTSource.removeFromMap: %s - %s", mSource, ex.getMessage()), ex);
            }
        }
    }

    public void addLayer(View childView, int childPosition) {
        if (!(childView instanceof AbstractSourceConsumer)) {
            return;
        }

        AbstractSourceConsumer layer = (AbstractSourceConsumer) childView;
        if (mMap == null) {
            mQueuedLayers.add(childPosition, layer);
        } else {
            addLayerToMap(layer, childPosition);
        }
    }

    public void removeLayer(int childPosition) {
        AbstractSourceConsumer layer;
        if (mQueuedLayers != null && mQueuedLayers.size() > 0) {
            layer = mQueuedLayers.get(childPosition);
        } else {
            layer = mLayers.get(childPosition);
        }
        removeLayerFromMap(layer, childPosition);
    }

    public AbstractSourceConsumer getLayerAt(int childPosition) {
        if (mQueuedLayers != null && mQueuedLayers.size() > 0) {
            return mQueuedLayers.get(childPosition);
        }
        return mLayers.get(childPosition);
    }

    protected void removeLayerFromMap(AbstractSourceConsumer layer, int childPosition) {
        if (mMapView != null && layer != null) {
            layer.removeFromMap(mMapView);
        }
        if (mQueuedLayers != null && mQueuedLayers.size() > 0) {
            mQueuedLayers.remove(childPosition);
        } else {
            mLayers.remove(childPosition);
        }
    }

    public abstract T makeSource();

    static public class OnPressEvent {
        public List<Feature> features;
        public LatLng latLng;
        public PointF screenPoint;

        public OnPressEvent(@NonNull List<Feature> features, @NonNull LatLng latLng, @NonNull PointF screenPoint) {
            this.features = features;
            this.latLng = latLng;
            this.screenPoint = screenPoint;
        }
    }

    public abstract void onPress(OnPressEvent event);

    public static boolean isDefaultSource(String sourceID) {
        return DEFAULT_ID.equals(sourceID);
    }
}

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTSource.kt
package com.mapbox.rctmgl.components.styles.sources

import android.content.Context
import com.mapbox.maps.extension.style.sources.getSource
import com.mapbox.maps.extension.style.sources.addSource
import com.mapbox.rctmgl.components.AbstractMapFeature
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView
import com.mapbox.maps.MapboxMap
import com.mapbox.rctmgl.components.styles.sources.AbstractSourceConsumer
import com.facebook.react.bridge.ReadableMap
import com.mapbox.rctmgl.components.styles.sources.RCTSource
import android.graphics.PointF
import android.view.View
import com.facebook.react.common.MapBuilder
import com.mapbox.geojson.Feature
import com.mapbox.maps.Style
import com.mapbox.maps.extension.style.StyleContract
import com.mapbox.maps.extension.style.sources.Source
import com.mapbox.rctmgl.components.styles.sources.RCTSource.OnPressEvent
import com.mapbox.rctmgl.utils.LatLng
import com.mapbox.rctmgl.utils.Logger
import java.lang.ClassCastException
import java.util.ArrayList
import java.util.HashMap

abstract class RCTSource<T : Source?>(context: Context?) : AbstractMapFeature(context) {
    @JvmField
    protected var mMapView: RCTMGLMapView? = null
    @JvmField
    protected var mMap: MapboxMap? = null
    var iD: String? = null
    @JvmField
    protected var mSource: T? = null
    protected var mHasPressListener = false
    protected var mTouchHitbox: Map<String, Double>? = null
    protected var mLayers: MutableList<AbstractSourceConsumer>
    private var mQueuedLayers: MutableList<AbstractSourceConsumer>?
    val layerIDs: List<String>
        get() {
            val layerIDs: MutableList<String> = ArrayList()
            for (i in mLayers.indices) {
                val layer = mLayers[i]
                val id = layer.iD
                if (id != null) {
                    layerIDs.add(id)
                }
            }
            return layerIDs
        }

    private fun getSourceAs(style: Style, id: String?): T? {
        val result = iD?.let { style.getSource(it) }
        return try {
            result as T?
        } catch (exception: ClassCastException) {
            null
        }
    }

    protected fun addLayerToMap(layer: AbstractSourceConsumer?, childPosition: Int) {
        if (mMapView == null || layer == null) {
            return
        }
        layer.addToMap(mMapView)
        if (!mLayers.contains(layer)) {
            mLayers.add(childPosition, layer)
        }
    }

    open fun hasPressListener(): Boolean {
        return mHasPressListener
    }

    fun setHasPressListener(hasPressListener: Boolean) {
        mHasPressListener = hasPressListener
    }

    fun setHitbox(map: ReadableMap) {
        val hitbox: MutableMap<String, Double> = HashMap()
        hitbox["width"] = map.getDouble("width")
        hitbox["height"] = map.getDouble("height")
        mTouchHitbox = hitbox
    }

    val touchHitbox: Map<String, Double>?
        get() {
            if (!hasPressListener()) {
                return null
            }
            return if (mTouchHitbox == null) {
                MapBuilder.builder<String, Double>()
                        .put("width", DEFAULT_HITBOX_WIDTH)
                        .put("height", DEFAULT_HITBOX_HEIGHT)
                        .build()
            } else mTouchHitbox
        }
    val layerCount: Int
        get() {
            var totalCount = 0
            if (mQueuedLayers != null) {
                totalCount = mQueuedLayers!!.size
            }
            totalCount += mLayers.size
            return totalCount
        }

    override fun addToMap(mapView: RCTMGLMapView) {
        mMapView = mapView
        mMap = mapView.getMapboxMap()
        mMap?.getStyle(object : Style.OnStyleLoaded {
            override fun onStyleLoaded(style: Style) {
                val existingSource = getSourceAs(style, iD)
                if (existingSource != null) {
                    mSource = existingSource
                } else {
                    mSource = makeSource()
                    style.addSource(mSource as StyleContract.StyleSourceExtension)
                }
                if (mQueuedLayers != null && mQueuedLayers!!.size > 0) { // first load
                    for (i in mQueuedLayers!!.indices) {
                        addLayerToMap(mQueuedLayers!![i], i)
                    }
                    mQueuedLayers = null
                } else if (mLayers.size > 0) { // handles the case of switching style url, but keeping layers on map
                    for (i in mLayers.indices) {
                        addLayerToMap(mLayers[i], i)
                    }
                }
            }
        })
    }

    override fun removeFromMap(mapView: RCTMGLMapView) {
        if (mLayers.size > 0) {
            for (i in mLayers.indices) {
                val layer = mLayers[i]
                layer.removeFromMap(mMapView)
            }
        }
        if (mQueuedLayers != null) {
            mQueuedLayers!!.clear()
        }
        if (mMap != null && mSource != null && mMap!!.getStyle() != null) {
            try {
                iD?.let { mMap?.getStyle()?.removeStyleSource(it) }
            } catch (ex: Throwable) {
                Logger.w(LOG_TAG, String.format("RCTSource.removeFromMap: %s - %s", mSource, ex.message), ex)
            }
        }
    }

    fun addLayer(childView: View?, childPosition: Int) {
        if (childView !is AbstractSourceConsumer) {
            return
        }
        val layer = childView
        if (mMap == null) {
            mQueuedLayers?.add(childPosition, layer)
        } else {
            addLayerToMap(layer, childPosition)
        }
    }

    fun removeLayer(childPosition: Int) {
        val layer: AbstractSourceConsumer
        layer = if (mQueuedLayers != null && mQueuedLayers!!.size > 0) {
            mQueuedLayers!![childPosition]
        } else {
            mLayers[childPosition]
        }
        removeLayerFromMap(layer, childPosition)
    }

    fun getLayerAt(childPosition: Int): AbstractSourceConsumer {
        return if (mQueuedLayers != null && mQueuedLayers!!.size > 0) {
            mQueuedLayers!![childPosition]
        } else mLayers[childPosition]
    }

    protected fun removeLayerFromMap(layer: AbstractSourceConsumer?, childPosition: Int) {
        if (mMapView != null && layer != null) {
            layer.removeFromMap(mMapView)
        }
        if (mQueuedLayers != null && mQueuedLayers!!.size > 0) {
            mQueuedLayers?.removeAt(childPosition)
        } else {
            mLayers.removeAt(childPosition)
        }
    }

    abstract fun makeSource(): T
    class OnPressEvent(var features: List<Feature>, var latLng: LatLng, var screenPoint: PointF)

    abstract fun onPress(event: OnPressEvent?)

    companion object {
        const val DEFAULT_ID = "composite"
        const val LOG_TAG = "RCTSource"
        const val DEFAULT_HITBOX_WIDTH = 44.0
        const val DEFAULT_HITBOX_HEIGHT = 44.0
        @JvmStatic
        fun isDefaultSource(sourceID: String): Boolean {
            return DEFAULT_ID == sourceID
        }
    }

    init {
        mLayers = ArrayList()
        mQueuedLayers = ArrayList()
    }
}